File: mnist_with_mlps/C++/MNIST_handwritten_digits_classification_with_MLPs.cpp
#include <torch/torch.h>
#include <torch/script.h>
#include <torch/data/datasets/mnist.h>
#include <torch/data/transforms.h>
#include <iostream>
#include <vector>

class Net : public torch::nn::Module {
public:
    Net() {
        conv1 = register_module("conv1", torch::nn::Conv2d(torch::nn::Conv2dOptions(1, 32, 3)));
        conv1_bn = register_module("conv1_bn", torch::nn::BatchNorm2d(32));
        conv2 = register_module("conv2", torch::nn::Conv2d(torch::nn::Conv2dOptions(32, 64, 3)));
        conv2_bn = register_module("conv2_bn", torch::nn::BatchNorm2d(64));
        conv3 = register_module("conv3", torch::nn::Conv2d(torch::nn::Conv2dOptions(64, 128, 3)));
        conv3_bn = register_module("conv3_bn", torch::nn::BatchNorm2d(128));
        fc1 = register_module("fc1", torch::nn::Linear(128 * 5 * 5, 256));
        fc1_bn = register_module("fc1_bn", torch::nn::BatchNorm1d(256));
        dropout = register_module("dropout", torch::nn::Dropout(0.5));
        fc2 = register_module("fc2", torch::nn::Linear(256, 128));
        fc2_bn = register_module("fc2_bn", torch::nn::BatchNorm1d(128));
        fc3 = register_module("fc3", torch::nn::Linear(128, 10));
    }

    torch::Tensor forward(torch::Tensor x) {
        x = conv1_bn->forward(conv1->forward(x));
        x = torch::relu(x);
        x = conv2_bn->forward(conv2->forward(x));
        x = torch::relu(x);
        x = torch::max_pool2d(x, 2);
        x = conv3_bn->forward(conv3->forward(x));
        x = torch::relu(x);
        x = torch::max_pool2d(x, 2);
        x = x.view({ -1, 128 * 5 * 5 });
        x = fc1_bn->forward(fc1->forward(x));
        x = torch::relu(x);
        x = dropout->forward(x);
        x = fc2_bn->forward(fc2->forward(x));
        x = torch::relu(x);
        x = dropout->forward(x);
        x = fc3->forward(x);
        return x;
    }

    torch::nn::Conv2d conv1{nullptr};
    torch::nn::BatchNorm2d conv1_bn{nullptr};
    torch::nn::Conv2d conv2{nullptr};
    torch::nn::BatchNorm2d conv2_bn{nullptr};
    torch::nn::Conv2d conv3{nullptr};
    torch::nn::BatchNorm2d conv3_bn{nullptr};
    torch::nn::Linear fc1{nullptr};
    torch::nn::BatchNorm1d fc1_bn{nullptr};
    torch::nn::Dropout dropout{nullptr};
    torch::nn::Linear fc2{nullptr};
    torch::nn::BatchNorm1d fc2_bn{nullptr};
    torch::nn::Linear fc3{nullptr};
};

void train(int epoch, int log_interval = 200) {
    model->train();

    for (const auto& batch : *train_loader) {
        auto data = batch.data.to(device);
        auto target = batch.target.to(device);

        optimizer.zero_grad();
        auto output = model->forward(data);
        auto loss = criterion->forward(output, target);
        loss.backward();
        optimizer.step();

        if (batch.index() % log_interval == 0) {
            std::cout << "Train Epoch: " << epoch << " ["
                << batch.index() * data.size(0) << "/"
                << train_loader->size().value() << " ("
                << 100. * batch.index() / train_loader->size().value()
                << "%)]\tLoss: " << loss.item<float>() << std::endl;
        }
    }
}

void validate(std::vector<float>& loss_vector, std::vector<float>& accuracy_vector) {
    model->eval();
    float val_loss = 0;
    int correct = 0;

    for (const auto& batch : *validation_loader) {
        auto data = batch.data.to(device);
        auto target = batch.target.to(device);
        auto output = model->forward(data);
        val_loss += criterion->forward(output, target).item<float>();
        auto pred = output.argmax(1); // get the index of the max log-probability
        correct += pred.eq(target).sum().item<int>();
    }

    val_loss /= validation_loader->size().value();
    loss_vector.push_back(val_loss);

    float accuracy = 100. * static_cast<float>(correct) / validation_loader->size().value();
    accuracy_vector.push_back(accuracy);

    std::cout << "\nValidation set: Average loss: " << val_loss
        << ", Accuracy: " << correct << "/" << validation_loader->size().value()
        << " (" << accuracy << "%)\n" << std::endl;
}

int main() {
    torch::manual_seed(1);

    torch::DeviceType device_type;
    if (torch::cuda::is_available()) {
        std::cout << "CUDA available. Using GPU." << std::endl;
        device_type = torch::kCUDA;
    }
    else {
        std::cout << "CUDA not available. Using CPU." << std::endl;
        device_type = torch::kCPU;
    }
    torch::Device device(device_type);

    const int64_t batch_size = 32;

    auto train_dataset = torch::data::datasets::MNIST(
        "./data",
        torch::data::datasets::MNIST::Mode::kTrain,
        torch::data::datasets::MNIST::kNone,
        torch::data::datasets::MNIST::kNone,
        torch::data::datasets::MNIST::kNone);
    auto train_loader = torch::data::make_data_loader(
        std::move(train_dataset), batch_size);

    auto validation_dataset = torch::data::datasets::MNIST(
        "./data",
        torch::data::datasets::MNIST::Mode::kTest,
        torch::data::datasets::MNIST::kNone,
        torch::data::datasets::MNIST::kNone,
        torch::data::datasets::MNIST::kNone);
    auto validation_loader = torch::data::make_data_loader(
        std::move(validation_dataset), batch_size);

    for (const auto& batch : *train_loader) {
        const auto& images = batch.data.to(device);
        const auto& labels = batch.target.to(device);
        std::cout << "images: " << images.sizes() << " type: "
            << images.scalar_type() << std::endl;
        std::cout << "labels: " << labels.sizes() << " type: "
            << labels.scalar_type() << std::endl;
        break;
    }

    Net model;
    model.to(device);

    torch::optim::SGD optimizer(
        model.parameters(), torch::optim::SGDOptions(0.01).momentum(0.5));
    torch::optim::StepLR scheduler(optimizer, 3, 0.1);
    torch::nn::CrossEntropyLoss criterion;

    std::cout << model << std::endl;

    int epochs = 10;
    std::vector<float> lossv, accv;

    for (int epoch = 1; epoch <= epochs; ++epoch) {
        train(epoch);
        validate(lossv, accv);
    }

    torch::save(model->state_dict(), "../../Srcs/mnist_trained_model_v3.pth");

    return 0;
}

File: mnist_with_mlps/python/MNIST_handwritten_digits_classification_with_MLPs.py
#%matplotlib inline

import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import datasets, transforms
import numpy as np
import matplotlib.pyplot as plt

if torch.cuda.is_available():
    device = torch.device('cuda')
else:
    device = torch.device('cpu')

print('Using PyTorch version:', torch.__version__, ' Device:', device)

batch_size = 32

train_dataset = datasets.MNIST('./data',
                               train=True,
                               download=True,
                               transform=transforms.ToTensor())
                               #transform=transform)

validation_dataset = datasets.MNIST('./data',
                                    train=False,
                                    transform=transforms.ToTensor())
                                    #transform=transform)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=batch_size,
                                           shuffle=True)

validation_loader = torch.utils.data.DataLoader(dataset=validation_dataset,
                                                batch_size=batch_size,
                                                shuffle=False)

for (images, labels) in train_loader:
    print('images:', images.size(), 'type:', images.type())
    print('labels:', labels.size(), 'type:', labels.type())
    break

pltsize=1
plt.figure(figsize=(10*pltsize, pltsize))

for i in range(10):
    plt.subplot(1,10,i+1)
    plt.axis('off')
    plt.imshow(images[i,:,:,:].numpy().reshape(28,28), cmap="gray_r")
    plt.title('Class: '+str(labels[i].item()))

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=3)
        self.conv1_bn = nn.BatchNorm2d(32)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3)
        self.conv2_bn = nn.BatchNorm2d(64)
        self.conv3 = nn.Conv2d(64, 128, kernel_size= 3)
        self.conv3_bn = nn.BatchNorm2d(128)
        self.fc1 = nn.Linear(128 * 5 * 5, 256)
        self.fc1_bn = nn.BatchNorm1d(256)
        self.dropout = nn.Dropout(0.5)
        self.fc2 = nn.Linear(256, 128)
        self.fc2_bn = nn.BatchNorm1d(128)
        self.dropout = nn.Dropout(0.5)
        self.fc3 = nn.Linear(128, 10)
        
    def forward(self, x):
        x = self.conv1_bn(self.conv1(x))
        x = nn.functional.relu(x)
        x = self.conv2_bn(self.conv2(x))
        x = nn.functional.relu(x)
        x = nn.functional.max_pool2d(x, 2)
        x = self.conv3_bn(self.conv3(x))
        x = nn.functional.relu(x)
        x = nn.functional.max_pool2d(x, 2)
        x = x.view(-1, 128 * 5 * 5)
        x = self.fc1_bn(self.fc1(x))
        x = nn.functional.relu(x)
        x = self.dropout(x)  
        x = self.fc2_bn(self.fc2(x))
        x = nn.functional.relu(x)
        x = self.dropout(x)  
        x = self.fc3(x);
        return x

model = Net().to(device)
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.5)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=3, gamma=0.1)
#scheduler.step()
criterion = nn.CrossEntropyLoss()

print(model)

def train(epoch, log_interval=200):
    # Set model to training mode
    model.train()

    # Loop over each batch from the training set
    for batch_idx, (data, target) in enumerate(train_loader):
        # Copy data to GPU if needed
        data = data.to(device)
        target = target.to(device)

        # Zero gradient buffers
        optimizer.zero_grad()

        # Pass data through the network
        output = model(data)

        # Calculate loss
        loss = criterion(output, target)

        # Backpropagate
        loss.backward()

        # Update weights
        optimizer.step()

        if batch_idx % log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.data.item()))

def validate(loss_vector, accuracy_vector):
    model.eval()
    val_loss, correct = 0, 0
    for data, target in validation_loader:
        data = data.to(device)
        target = target.to(device)
        output = model(data)
        val_loss += criterion(output, target).data.item()
        pred = output.data.max(1)[1] # get the index of the max log-probability
        correct += pred.eq(target.data).cpu().sum()

    val_loss /= len(validation_loader)
    loss_vector.append(val_loss)

    accuracy = 100. * correct.to(torch.float32) / len(validation_loader.dataset)
    accuracy_vector.append(accuracy)

    print('\nValidation set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        val_loss, correct, len(validation_loader.dataset), accuracy))

#%%time
epochs = 10

lossv, accv = [], []
for epoch in range(1, epochs + 1):
    train(epoch)
    validate(lossv, accv)

torch.save(model.state_dict(), '../../Srcs/mnist trained model v3.pth')

plt.show();

plt.figure(figsize=(5,3))
plt.plot(np.arange(1,epochs+1), lossv)
plt.title('validation loss')
plt.show();

plt.figure(figsize=(5,3))
plt.plot(np.arange(1,epochs+1), accv)
plt.title('validation accuracy');
plt.show();


